doc/user/content/sql/create-sink/kafka.md

```diff
@@ -67,6 +67,7 @@ Field                               | Value  | Description
 `TOPIC`                             | `text`              | The name of the Kafka topic to write to.
 `COMPRESSION TYPE`                  | `text`              | The type of compression to apply to messages before they are sent to Kafka: `none`, `gzip`, `snappy`, `lz4`, or `zstd`.<br>Default: `none`.
 `TRANSACTIONAL ID PREFIX`           | `text`              | The prefix of the transactional ID to use when producing to the Kafka topic.<br>Default: `materialize-{REGION ID}-{CONNECTION ID}-{SINK ID}`.
+`PARTITION HASHER`                  | expression          | A SQL expression returning a hash that can be used for partition assignment. See [Partitioning](#partitioning) for details.
```

Comment:
How much do we need to bear the weight of choosing the hash function on the user? Could we expose this as `PARTITION KEY` with an additional `PARTITION HASH TYPE` that defaults to one of the hash functions?

Reply:
Yeah, it's a fair question. My philosophy was that we should build the most powerful and flexible thing for users first, and then if common patterns emerge, we can wrap them up in syntax sugar.

The reason I think it's useful to have full control over the partition strategy is so that you can exactly match the way your existing code partitions things. This is particularly relevant for composite keys. Suppose you want to hash both customer and order ID for the partition. How do you combine those two values together? Maybe your code was previously separating those two quantities with a `|` character:

```
murmur2(customer_id::text || '|' || order_id::text)
```

The design here lets you exactly match the way your code was already working.

If we were to offer a higher level interface like:

```
PARTITION KEY = (customer_id, order_id)
PARTITION HASH FUNCTION = murmur2
```

now there's a hard question to answer about how to combine those fields. Do we use the same format that you specified for the key? But hashing Avro/JSON values is fraught for all sorts of reason.

I do think we'll want to expose that higher level interface some day, but I don't think we'll have a clear answer to the question of how to hash composite partition keys until we see how the `PARTITION HASHER` is used in the wild.

Reply:
Gotcha.